### PR TITLE
Univariate Poly structure redo

### PR DESCRIPTION
**Pull request by srajangarg**
Currently, the `UnivariateIntPoly` strcuture and functions did not match the type of functions as used in rest of SymEngine. Fixed that.
Mainly
`func(classT &a, classT &b)` to
`func(RCP<classT> &a, RCP<classT> &b)`

Also, had a question about including `add_poly` as a part of `add`. I have done so in my PR, but calling add on two polys will return a `RCP<Basic>` and this will have to be casted to `RCP<UniPoly>`, if further poly functions are to use it. I was wondering if we could overload `add()` so as to return `RCP<UniPoly>` directly? 

@Sumith1896 @isuruf 

Edit : major changes to ensue

### CODE REVIEW

File: symengine/polynomial.cpp [hunk 1]
```diff
@@ -208,34 +208,34 @@ bool UnivariateIntPolynomial::is_pow() const {
     return false;
 }
 
-RCP<const UnivariateIntPolynomial> add_poly(const UnivariateIntPolynomial &a, const UnivariateIntPolynomial &b) {
+RCP<const UnivariateIntPolynomial> add_poly(const RCP<const UnivariateIntPolynomial> &a, const RCP<const UnivariateIntPolynomial> &b) {
```
Comment (Sumith1896) [inline symengine/polynomial.cpp]: The exact reason for this is here https://github.com/symengine/symengine/pull/406#issuecomment-77592690

Sorry I couldn't find the exact comment earlier.

**Comment by Sumith1896** [comment]
You could also add `mul_poly` to `Mul`

File: symengine/add.cpp [hunk 2]
```diff
@@ -5,7 +5,9 @@
 #include <symengine/pow.h>
 #include <symengine/complex.h>
 #include <symengine/functions.h>
+#include <symengine/polynomial.h>
 
+using SymEngine::is_a;
```
Comment (Sumith1896) [inline symengine/add.cpp]: Is this needed? There are other instances(in the same file too) where `is_a<Mul>` etc., are used without `SymEngine::`

**Comment by srajangarg** [comment]
@Sumith1896 Okay.
What do you think about overloading `add` as asked earlier?

**Comment by isuruf** [comment]
What happens when polys and other symbolics are mixed. 
```
a = sin(y) + poly(x**2, x)
b = a + x
```

**Comment by srajangarg** [comment]
Nothing in particular. An `Add` is formed,

**Comment by isuruf** [comment]
Yes. that should be fixed. The design has to be thought out carefully for seamless operation between polynomials and symbolics.

**Comment by srajangarg** [comment]
I'm not sure about what you mean. Do we need a `poly_to_basic()` function? and vice versa?

**Comment by isuruf** [comment]
No. See what SymPy does. You can work with Polynomials and symbolics easily. `Poly(x**2, x) + x` is converted to `Poly(x**2+x, x)`

**Comment by srajangarg** [comment]
Oh okay, will be done

**Comment by isuruf** [comment]
That's fine. If you want the polynomial methods you can always use `add_poly` methods

**Pull request closed without merging**